src/ray/object_manager/pull_manager.cc

```diff
@@ -90,23 +94,45 @@ void PullManager::TryPull(const ObjectID &object_id) {
     return;
   }
 
-  // Choose a random client to pull the object from.
-  // Generate a random index.
-  std::uniform_int_distribution<int> distribution(0, node_vector.size() - 1);
-  int node_index = distribution(gen_);
-  NodeID node_id = node_vector[node_index];
-  // If the object manager somehow ended up choosing itself, choose a different
-  // object manager.
-  if (node_id == self_node_id_) {
-    std::swap(node_vector[node_index], node_vector[node_vector.size() - 1]);
-    node_vector.pop_back();
-    RAY_LOG(WARNING)
-        << "The object manager with ID " << self_node_id_ << " is trying to pull object "
-        << object_id << " but the object table suggests that this object manager "
-        << "already has the object. It is most likely due to memory pressure, object "
-        << "pull has been requested before object location is updated.";
-    node_id = node_vector[node_index % node_vector.size()];
-    RAY_CHECK(node_id != self_node_id_);
+  NodeID node_id;
+  int num_candidates = 0;
+  for (NodeID potential : node_vector) {
+    if (inflight_pulls.count(potential) > 0) {
+      continue;
+    }
+    num_candidates++;
+    std::uniform_int_distribution<int> distribution(0, num_candidates - 1);
+    if (distribution(gen_) == 0) {
+      node_id = potential;
```

Comment:
Can you write a comment what this block is exactly is doing? 

Q: Not familiar with how this works, but is there any possibility this if statement is always false and the node id is not set when it is supposed to?

Reply:
Yes, that's the case where the node id remains nil.